### PR TITLE
Worker and ProcessingManager implementation

### PR DESCRIPTION
**Pull request by lgfrbcsgo**
~**DON'T EVEN THINK ABOUT MERGING YET!**~

- Added `Worker` class as a wrapper for `WetManager`, `WebPageDigester`, `LanguageExtractor` and `Storer`

- Added `ProcessingManager` class for managing WET CC file paths and inter process communication

These changes **are** tested ~since I am not able to update the submodule and the DB isn't in place yet.
I've only opened this pull request as a discussion platform!~

~**DON'T EVEN THINK ABOUT MERGING YET!**~

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -1,11 +1,46 @@
...
+            return;
+        }
+
+        // I don't think we need more than one Worker as the cluster will probably call 'workOn()' in parallel
+        const worker = new Worker(entities, false, ["en"]);
+
+        // start threads
+        let i = 0;
+        while (i < numCPUs && ccWetPaths.length > 0) {
+            cluster.fork({worker: worker, wetPath: ccWetPaths.pop()});
```
Comment (lgfrbcsgo) [inline src/app.ts]: Well, sharing objects won't work... 
http://stackoverflow.com/questions/14826349/how-to-keep-variables-that-share-all-node-processes-in-node-cluster

File: src/app.ts [hunk 2]
```diff
@@ -1,11 +1,4 @@
...
-        console.log("SHIT!\n", err);
-        return;
-    }
-    console.log("\n\n\n\n");
-    console.log("entities:\n", entities);
-
-});
\ No newline at end of file
+// fingers crossed, hope this works :3
+ProcessingManager.run();
```
Comment (lgfrbcsgo) [inline src/app.ts]: @nyxathid can you please test this as you're the only one with a working dummy DB.

File: src/worker.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
+import ReadableStream = NodeJS.ReadableStream;
+import EventEmitter = NodeJS.EventEmitter;
```
Comment (nyxathid) [inline src/worker.ts]: error in compiler `worker.js`: `ReferenceError: NodeJS is not defined` o_O

Comment (lgfrbcsgo) [inline src/worker.ts]: That's very strange... The Travis build passed which runs `tsc`... Maybe delete the `out` folder and run `tsc` again?

File: src/worker.ts [hunk 4]
```diff
@@ -50,57 +46,91 @@ export class Worker extends EventEmitter {
...
+        let streamFinished = false;
+        let pendingPages = 0;
+
+        /**
+         * Gets called once the unpacked WET stream starts and pipes stream to WARC parser
+         * Parses WET entries and then calls for each of them 'onWARCdata'
+         * @param err
+         * @param response
+         */
+        let onFileStreamReady = (err? : Error, response? : ReadableStream) => {
```
Comment (nyxathid) [inline src/worker.ts]: Do variables (instead functions) solve weird stuff with `this`?

File: src/worker.ts [hunk 5]
```diff
@@ -31,6 +32,7 @@ export class Worker extends EventEmitter {
 
         this.caching = caching || process.env.caching || false;
         this.languageCodes = languageCodes || process.env.languageCodes;
+        this.processID = process.pid;
```
Comment (nyxathid) [inline src/worker.ts]: cool :)

**Comment by nyxathid** [comment]
Let's test everything tomorrow during the meeting! Especially how the workers behave when they are done with a single WET file.

**Comment by nyxathid** [comment]
merged master into this branch

**Comment by nyxathid** [comment]
Tested: 
- get data from (local) DB [OK]
- search for entities [OK]
- save to azure [OK]
- download link works [OK] 
- store metadata into (local) DB [NOT OK] // but it might be because of my DB (entities are missing)
 
looks pretty good to me, can someone merge into develop?

File: src/worker.ts (hunk 4 above)
Comment (lgfrbcsgo) [inline src/worker.ts]: Yes, lambda functions in ECMA6 perserve the `this` context.

**Comment by lgfrbcsgo** [comment]
We're almost done here... The call to `CCPathLoader` is missing in `ProcessingManager`!

**Comment by lgfrbcsgo** [comment]
Also wasn't there an issue with case sensitivity?

**Comment by nyxathid** [comment]
Yes, next PR is fine. Otherwise we can rename this branch to `develop` ^^

File: src/processing-manager.ts [hunk 6]
```diff
@@ -60,11 +61,7 @@ export class ProcessingManager {
             }
         };
 
-        // TODO: programmatically add CC paths to array
-        onPaths(undefined,
-            ['crawl-data/CC-MAIN-2017-09/segments/1487501172017.60/wet/' +
-            'CC-MAIN-20170219104612-00150-ip-10-171-10-108.ec2.internal.warc.wet.gz']
-        );
+        CCPathLoader.loadPaths(undefined, onPaths);
```
Comment (nyxathid) [inline src/processing-manager.ts]: Maybe we want a subset of those?

Comment (lgfrbcsgo) [inline src/processing-manager.ts]: That should be done in the CLI PR. This is better than a single link though for a start 😎

**Pull request merged**